laos_gggi/GPCC_data_loader.py

```diff
@@ -2,81 +2,111 @@
+        with ZipFile(
+            output_path + "/shapefiles/lao_adm_ngd_20191112_shp.zip", "r"
+        ) as zObject:
+            zObject.extractall(path=output_path + "/shapefiles")
+
+    # Verify if the gpcc files are extracted (Note:gzip.open does not support loops)
     if not exists(path_to_GPCC_unzipped):
-        with gzip.open(output_path + "/gpcc/gpcc_raw_1981_1990.nc.gz", 'rb') as f_in:
-            with open(output_path + "/gpcc/gpcc_raw_1981_1990.nc", 'wb') as f_out:
+        with gzip.open(output_path + "/gpcc/gpcc_raw_1981_1990.nc.gz", "rb") as f_in:
+            with open(output_path + "/gpcc/gpcc_raw_1981_1990.nc", "wb") as f_out:
                 shutil.copyfileobj(f_in, f_out)
 
-        with gzip.open(output_path + "/gpcc/gpcc_raw_1991_2000.nc.gz", 'rb') as f_in:
-            with open(output_path + "/gpcc/gpcc_raw_1991_2000.nc", 'wb') as f_out:
+        with gzip.open(output_path + "/gpcc/gpcc_raw_1991_2000.nc.gz", "rb") as f_in:
+            with open(output_path + "/gpcc/gpcc_raw_1991_2000.nc", "wb") as f_out:
                 shutil.copyfileobj(f_in, f_out)
 
-        with gzip.open(output_path + "/gpcc/gpcc_raw_2001_2010.nc.gz", 'rb') as f_in:
-            with open(output_path + "/gpcc/gpcc_raw_2001_2010.nc", 'wb') as f_out:
+        with gzip.open(output_path + "/gpcc/gpcc_raw_2001_2010.nc.gz", "rb") as f_in:
+            with open(output_path + "/gpcc/gpcc_raw_2001_2010.nc", "wb") as f_out:
                 shutil.copyfileobj(f_in, f_out)
 
-        with gzip.open(output_path + "/gpcc/gpcc_raw_2011_2020.nc.gz", 'rb') as f_in:
-            with open(output_path + "/gpcc/gpcc_raw_2011_2020.nc", 'wb') as f_out:
+        with gzip.open(output_path + "/gpcc/gpcc_raw_2011_2020.nc.gz", "rb") as f_in:
+            with open(output_path + "/gpcc/gpcc_raw_2011_2020.nc", "wb") as f_out:
                 shutil.copyfileobj(f_in, f_out)
 
     if not exists(gpcc_processed_path):
-        # Import the world shapefile 
-        world_shapefile = geo.read_file(world_shapefile_path).rename(columns = {"ISO_A3": "country_code","FORMAL_EN":"country",
-                                                                                'CONTINENT': "continent", 'REGION_UN': "region" })
-        #Open gpcc files, transform them to geopandas shapefile geometry and merge them with the world shapefiles
+        # Import the world shapefile
+        world_shapefile = geo.read_file(world_shapefile_path).rename(  # noqa
+            columns={
+                "ISO_A3": "country_code",
+                "FORMAL_EN": "country",
+                "CONTINENT": "continent",
+                "REGION_UN": "region",
+            }
+        )
+        # Open gpcc files, transform them to geopandas shapefile geometry and merge them with the world shapefiles
         for x in list(path_to_GPCC_raw.keys()):
-            exec(f'data_{x} = xr.open_dataset(output_path + "/gpcc/gpcc_raw_" + x + ".nc")')
+            exec(
```

Comment:
Using `exec` should be avoided in general, and it doesn't seem necessary here. You can build up your results incrementally instead:

```
result_df = None
for key in path_to_GPCC_raw:
    data = xr.open_dataset(output_path + "/gpcc/gpcc_raw_" + x + ".nc")'
    df = data_{x}["precip"].to_dataframe().reset_index()
    df_gep = geo.GeoDataFrame(df_{x}, geometry=geo.points_from_xy(df_{x}["lat"], df_{x}["lon"]), crs="EPSG:4326")'

   # ...

   result_df = df if result_df is None else pd.concat([result_df, df], axis=0)

(result_df
    .reset_index()
    .pivot_table(values="precip", index=["country_code", "time"], aggfunc="mean")
    .to_csv(gpcc_processed_path)
)
```

Reply:
Understood!

Done!